src/videodownlinkbandwidthpolicy/VideoPreferences.ts

```diff
@@ -0,0 +1,234 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import Eq from './Eq';
+import PartialOrd from './PartialOrd';
+import VideoPreference from './VideoPreference';
+
+class ObjectSet<T extends Eq & PartialOrd> implements Iterable<T> {
+  constructor(private items: T[] = []) {}
+
+  static default(): ObjectSet<VideoPreference> {
+    return new ObjectSet([]);
+  }
+
+  // Returns the items in sorted order.
+  [Symbol.iterator](): Iterator<T, T, undefined> {
+    let i = 0;
+    const items = this.items;
+    return {
+      next(): IteratorResult<T> {
+        if (i < items.length) {
+          return {
+            done: false,
+            value: items[i++],
+          };
+        }
+        return {
+          done: true,
+          value: null,
+        };
+      },
+    };
+  }
+
+  first(): T | undefined {
+    return this.items[0];
+  }
+
+  add(item: T): void {
+    this.items.push(item);
+  }
+
+  replaceFirst(newItem: T, f: (item: T) => boolean): void {
+    this.items = this.items.filter(a => !f(a));
+    this.add(newItem);
+  }
+
+  remove(item: T): void {
+    this.items = this.items.filter(a => !a.equals(item));
+  }
+
+  clear(): void {
+    this.items = [];
+  }
+
+  isEmpty(): boolean {
+    return this.items.length === 0;
+  }
+
+  equals(other: this): boolean {
+    if (other === this) {
+      return true;
+    }
+
+    if (other.items.length !== this.items.length) {
+      return false;
+    }
+
+    // TODO: if we keep the arrays sorted correctly, not just by priority, then
+    // we don't need to do this painstaking O(n^2) work.
+    for (const item of this.items) {
+      if (!other.items.some(a => a.equals(item))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  has(item: T): boolean {
+    return this.items.some(a => a.equals(item));
+  }
+
+  some(f: (item: T) => boolean): boolean {
+    return this.items.some(f);
+  }
+
+  clone(): ObjectSet<T> {
+    return new ObjectSet([...this.items]);
+  }
+
+  sort(): void {
+    this.items.sort((a, b) => a.partialCompare(b));
+  }
+
+  modify(): SetBuilder<T> {
+    // COW.
+    return new SetBuilder(this);
+  }
+}
+
+class SetBuilder<T extends Eq & PartialOrd> {
+  private copied = false;
+
+  constructor(private items: ObjectSet<T> = new ObjectSet()) {}
+
+  private cow(): void {
+    if (this.copied) {
+      return;
+    }
+    this.items = this.items.clone();
+    this.copied = true;
+  }
+
+  add(item: T): void {
+    // Don't actually need to COW unless the item is there to add.
+    if (this.items.has(item)) {
+      return;
+    }
+    this.cow();
+    this.items.add(item);
+  }
+
+  replaceFirst(newItem: T, f: (item: T) => boolean): void {
+    // Don't actually need to COW unless the item is there to add
+    // or the previous is there to remove
+    if (this.items.has(newItem) && !this.items.some(f)) {
```

Comment:
Careful! 

What's the outcome of

```
[1, 2, 3, 4].replaceFirst(1, isEven);
```

If it's 

```
[1, 3, 4]
```

then you can't do this check as written.

Reply:
Hmm, i think this is fine?  I did rewrite the comment since before it describes `!(!this.items.has(newItem) || this.items.some(f))` which is a bit more verbose then was used.

In the above case `!this.items.some(f) === false` so it won't enter the branch.  I added a handful more unit tests, including the one you mention.

Reply:
I think as written you'll get

```
[1, 2, 3, 4]
```

so the intent to replace the first even item is not fulfilled because `1` is already present in the array.

Reply:
Since its `&&` though (again I think maybe comment caused confusion because i used `or`) it won't hit the branch.  I confirmed with the following test I added (shouldn't have posted comments before pushing, `build:release` always takes longer then i expect :P) :
```
  it('replaces only first occurance', () => {
    const b1: MutableVideoPreferences = VideoPreferences.prepare();
    b1.add(new VideoPreference('foo', 1));
    b1.add(new VideoPreference('foo', 2));
    b1.add(new VideoPreference('foo', 3));
    b1.add(new VideoPreference('foo', 4));
    b1.replaceFirst(new VideoPreference('foo', 1), (item: VideoPreference) => {
      return item.priority % 2 === 0;
    });
    const p1 = b1.build();

    const b2 = VideoPreferences.prepare();
    b2.add(new VideoPreference('foo', 1));
    b2.add(new VideoPreference('foo', 3));
    b2.add(new VideoPreference('foo', 4));
    const p2 = b2.build();

    expect(p1.equals(p2)).to.be.true;
  });
 ```